src/ios/ListLibraryItems.m

```diff
@@ -353,8 +356,56 @@ - (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)data
         }
     }
 }
+
 - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task needNewBodyStream:(nonnull void (^)(NSInputStream * _Nullable))completionHandler {
     NSLog(@"needNewBodyStream");
 }
 
+#pragma mark - Timeout check methods
+// This function starts a repeating timer. As long as it won't be stopped, the timer will call each 30 seconds a function.
+- (void)startCheckingProgress:(id)sender {
+    if(!uploadTimeout){
+        // Initializing variables
+        oldProgressUploadData = newProgressUploadData = 0;
+        // Creating timeout and starting it, passing through userInfo the old progress and new progress.
+        uploadTimeout = [NSTimer scheduledTimerWithTimeInterval:30
+                                                         target:self
+                                                       selector:@selector(checkProgress:)
+                                                       userInfo:nil
+                                                        repeats:YES];
+        NSLog(@"Timeout started !");
+    }
+}
+
+// This function stops the repeating timer, it is useful when an error occured during upload for example.
+- (void)stopCheckingProgress:(id)sender {
+    if([uploadTimeout isValid]) {
+        [uploadTimeout invalidate];
+    }
+    uploadTimeout = nil;
+    NSLog(@"Timeout stopped !");
+}
+
+// This function is called each 30 seconds from the timer when it is active. It calls hasUploadedNewData to know if there is new data since the last 30 seconds.
+- (void)checkProgress:(NSTimer *)timeout {
+    if(![self hasUploadedNewData:oldProgressUploadData :newProgressUploadData]) {
+        [currentTask cancel];
+        NSLog(@"Task canceled");
+        [self performSelectorOnMainThread:@selector(stopCheckingProgress:) withObject:nil waitUntilDone:YES];
+    } else {
+        oldProgressUploadData = newProgressUploadData;
+    }
+}
+
+// This function checks if the old value equals to the new value.
+- (bool)hasUploadedNewData:(int64_t)oldData :(int64_t)newData {
+    if(oldData == newData) {
+        NSLog(@"No data send since 30 seconds");
+        return false;
+    } else {
+        NSLog(@"New data");
```

Comment:
Do we really want to keep this log? I assume this would create a lot of noise in the console.